### PR TITLE
PruningVarExpand should handle incoming null nodes

### LINKED ISSUE #9492: OPTIONAL MATCH throws NullPointerException (Neo4j 3.2.1)
When I switch from Neo4j 3.0.7 to 3.2.1 one of my queries crashes with NullPointerException in
org.neo4j.cypher.internal.spi.v3_2.TransactionBoundQueryContext.getRelationshipsForIds.
With 3.0.7 the same query is running fine.

**Neo4j Version:** 3.2.1
**Operating System:** Windows 10
**API:** Cypher

### Steps to reproduce

- start data base and empty it
- create some nodes that will show the error
```
  CREATE (cust :DoorsObject {eid: 1})
  CREATE (cust) <-[:Links]- (req :DoorsObject {eid: 1})
  CREATE (reqM :DoorsFormalModule {eid: 1}) -[:Contains]-> (req)
```
- create some nodes that are OK
```
  CREATE (cust :DoorsObject {eid: 2})
  CREATE (cust) <-[:Links]- (req :DoorsObject {eid: 2})
  CREATE (reqM :DoorsFormalModule {eid: 2}) -[:Contains]-> (req)
  CREATE (req) <-[:Links]- (:Link) <-[:Links]- (tst :DoorsObject {eid: 2})
  CREATE (tstM :DoorsFormalModule {eid: 2}) -[:Contains]-> (tst)
```
- run the query
```
  MATCH
      (cust :DoorsObject)
  WITH cust
  OPTIONAL MATCH
    (cust)
       <-[:Links *1..]-
    (req :DoorsObject)
       <-[:Contains]-
    (reqM :DoorsFormalModule)
  WITH DISTINCT cust, reqM, req
  // WHERE req IS NOT NULL // Uncomment so 3.2.1 doesn't crash
  OPTIONAL MATCH
    (req)
      <-[:Links *2]-
    (tst :DoorsObject)
      <-[:Contains]-
    (tstM :DoorsFormalModule)
  WITH cust, reqM, req, tstM, tst
  // WHERE tstM IS NOT NULL
  RETURN DISTINCT
    cust.eid,
    reqM.eid,
    tstM.eid
```

### Expected behavior

Neo4j 3.0.7 returns:
```
╒════════╤════════╤════════╕
│cust.eid│reqM.eid│tstM.eid│
╞════════╪════════╪════════╡
│1       │1       │(null)  │
├────────┼────────┼────────┤
│1       │(null)  │(null)  │
├────────┼────────┼────────┤
│2       │2       │2       │
├────────┼────────┼────────┤
│2       │2       │(null)  │
├────────┼────────┼────────┤
│2       │(null)  │(null)  │
└────────┴────────┴────────┘
```

### Actual behavior

Neo4j 3.2.1 crashes with
```
java.lang.NullPointerException: null
 at org.neo4j.cypher.internal.spi.v3_2.TransactionBoundQueryContext.getRelationshipsForIds (TransactionBoundQueryContext.scala:134)
    org.neo4j.cypher.internal.compatibility.v3_2.ExceptionTranslatingQueryContext$$anonfun$getRelationshipsForIds$1.apply (ExceptionTranslatingQueryContext.scala:195)
    org.neo4j.cypher.internal.compatibility.v3_2.ExceptionTranslatingQueryContext$$anonfun$getRelationshipsForIds$1.apply (ExceptionTranslatingQueryContext.scala:195)
    org.neo4j.cypher.internal.spi.v3_2.ExceptionTranslationSupport$class.translateException (ExceptionTranslationSupport.scala:32)
    org.neo4j.cypher.internal.compatibility.v3_2.ExceptionTranslatingQueryContext.translateException (ExceptionTranslatingQueryContext.scala:34)
    org.neo4j.cypher.internal.compatibility.v3_2.ExceptionTranslatingQueryContext.getRelationshipsForIds (ExceptionTranslatingQueryContext.scala:195)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PruningVarLengthExpandPipe$Expandable$class.expand (PruningVarLengthExpandPipe.scala:300)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PruningVarLengthExpandPipe$PrePruningDFS.expand (PruningVarLengthExpandPipe.scala:81)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PruningVarLengthExpandPipe$PrePruningDFS.next (PruningVarLengthExpandPipe.scala:98)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PruningVarLengthExpandPipe$LoadNext.next (PruningVarLengthExpandPipe.scala:270)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PruningVarLengthExpandPipe$$anon$1.<init> (PruningVarLengthExpandPipe.scala:340)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PruningVarLengthExpandPipe.internalCreateResults (PruningVarLengthExpandPipe.scala:338)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PipeWithSource.createResults (Pipe.scala:82)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PipeWithSource.createResults (Pipe.scala:79)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PipeWithSource.createResults (Pipe.scala:79)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PipeWithSource.createResults (Pipe.scala:79)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.PipeWithSource.createResults (Pipe.scala:79)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.ApplyPipe$$anonfun$internalCreateResults$1.apply (ApplyPipe.scala:33)
    org.neo4j.cypher.internal.compiler.v3_2.pipes.ApplyPipe$$anonfun$internalCreateResults$1.apply (ApplyPipe.scala:30)
    scala.collection.Iterator$$anon$12.nextCur (Iterator.scala:434)
    scala.collection.Iterator$$anon$12.hasNext (Iterator.scala:440)
    scala.collection.Iterator$$anon$11.hasNext (Iterator.scala:408)
    scala.collection.Iterator$$anon$13.hasNext (Iterator.scala:461)
    scala.collection.Iterator$$anon$11.hasNext (Iterator.scala:408)
    org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator$$anonfun$hasNext$1.apply$mcZ$sp (ResultIterator.scala:60)
    org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator$$anonfun$hasNext$1.apply (ResultIterator.scala:58)
    org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator$$anonfun$hasNext$1.apply (ResultIterator.scala:58)
    org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator$$anonfun$failIfThrows$1.apply (ResultIterator.scala:94)
    org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator.decoratedCypherException (ResultIterator.scala:103)
    org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator.failIfThrows (ResultIterator.scala:92)
    org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator.hasNext (ResultIterator.scala:57)
    org.neo4j.cypher.internal.compiler.v3_2.PipeExecutionResult.hasNext (PipeExecutionResult.scala:77)
    org.neo4j.cypher.internal.compatibility.v3_2.ExecutionResultWrapper.hasNext (ExecutionResultWrapper.scala:91)
    org.neo4j.cypher.internal.compatibility.ClosingExecutionResult.org$neo4j$cypher$internal$compatibility$ClosingExecutionResult$$closeIfEmpty (ClosingExecutionResult.scala:178)
    org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anon$1$$anonfun$next$1.apply (ClosingExecutionResult.scala:63)
    org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anon$1$$anonfun$next$1.apply (ClosingExecutionResult.scala:61)
    org.neo4j.cypher.internal.compatibility.v3_2.exceptionHandler$runSafely$.apply (exceptionHandler.scala:90)
    org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anon$1.next (ClosingExecutionResult.scala:61)
    org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anon$1.next (ClosingExecutionResult.scala:55)
    org.neo4j.cypher.internal.javacompat.ExecutionResult.next (ExecutionResult.java:242)
    org.neo4j.cypher.internal.javacompat.ExecutionResult.next (ExecutionResult.java:54)
    clojure.lang.RT$4.invoke (RT.java:509)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.RT.seq (RT.java:521)
    clojure.core$seq__4357.invokeStatic (core.clj:137)
    clojure.core/seq (core.clj:137)
    com.bosch.as.metrics.graph.manager$print_execute$fn__1234.invoke (manager.clj:77)
    com.bosch.as.metrics.graph.manager$print_execute.invokeStatic (manager.clj:72)
    com.bosch.as.metrics.graph.manager$print_execute.invoke (manager.clj:70)
    com.bosch.as.metrics.adapter.birt.data_set$CustSpecs_TestResults.invokeStatic (data_set.clj:64)
    com.bosch.as.metrics.adapter.birt.data_set$CustSpecs_TestResults.invoke (data_set.clj:62)
    user$eval9670.invokeStatic (NO_SOURCE_FILE:4)
    user$eval9670.invoke (NO_SOURCE_FILE:4)
    clojure.lang.Compiler.eval (Compiler.java:6927)
    clojure.lang.Compiler.eval (Compiler.java:6890)
    clojure.core$eval.invokeStatic (core.clj:3105)
    clojure.core$eval.invoke (core.clj:3101)
    clojure.main$repl$read_eval_print__7408$fn__7411.invoke (main.clj:240)
    clojure.main$repl$read_eval_print__7408.invoke (main.clj:240)
    clojure.main$repl$fn__7417.invoke (main.clj:258)
    clojure.main$repl.invokeStatic (main.clj:258)
    clojure.main$repl_opt.invokeStatic (main.clj:322)
    clojure.main$repl_opt.invoke (main.clj:318)
    clojure.main$main.invokeStatic (main.clj:421)
    clojure.main$main.doInvoke (main.clj:384)
    clojure.lang.RestFn.invoke (RestFn.java:436)
    clojure.lang.Var.invoke (Var.java:388)
    clojure.lang.AFn.applyToHelper (AFn.java:160)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.main.main (main.java:37)
```

### PR DESCRIPTION
**Pull request by Lojjs**
Should be null forward merged, because 3.3 has a separate fix for the same bug (https://github.com/neo4j/neo4j/pull/10367) 

Changelog: Fixes #9492, so that PruningVarExpand after optional match will handle incoming null nodes

### CODE REVIEW

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/pipes/PruningVarLengthExpandPipe.scala [hunk 1]
```diff
@@ -260,14 +260,19 @@ case class PruningVarLengthExpandPipe(source: Pipe,
...
+        if (node == null) {
+          (Empty, null)
+        } else {
+          val nextState = new PrePruningDFS(whenEmptied = this,
+            node = node,
+            path = new Array[Long](max),
+            pathLength = 0,
+            state = state,
+            row = row,
+            expandMap = Primitive.longObjectMap[FullExpandDepths]())
```
Comment (fickludd) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/pipes/PruningVarLengthExpandPipe.scala]: To me this indentation is slightly less readable, but only change it if you agree :).

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 2]
```diff
@@ -752,4 +752,69 @@ return p""")
...
+        |  WITH cust
+        |  OPTIONAL MATCH (cust)<-[:Links *1..]-(req :DoorsObject)<-[:Contains]-(reqM :DoorsFormalModule)
+        |  WITH DISTINCT cust, reqM, req
+        |  OPTIONAL MATCH (req)<-[:Links *2]-(tst :DoorsObject)<-[:Contains]-(tstM :DoorsFormalModule)
+        |  WITH cust, reqM, req, tstM, tst
+        |  RETURN DISTINCT
+        |    cust.eid,
+        |    reqM.eid,
+        |    tstM.eid
+      """.stripMargin
```
Comment (fickludd) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Does this test trigger a bug which is not caught by the simpler test above?

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 3]
```diff
@@ -752,4 +752,69 @@ return p""")
...
+
+  test("Should handle optional match with null parts and distinct without NullPointerException") {
+
+    graph.execute("CREATE (:X {eid: 1})")
+
+    val query =
+      """
+        |  MATCH (x:X)
+        |  WITH x
+        |  OPTIONAL MATCH (x)<-[:L]-(req)
```
Comment (fickludd) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Is the expand needed here, our could we simply `OPTIONAL MATCH (req:NON_EXISTING_LABEL)`?

**Review by fickludd — APPROVED** [review]
Got some minor comments, but nothing that necessarily needs to be fixed. Good to have this fixed.

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 2 above)
Comment (Lojjs) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Not really, it's the same bug but this test is the code from the Github issue, while the simpler test is the simplest query we could found that generated the same error, so it's more for understanding.

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 3 above)
Comment (Lojjs) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Don't really understand what you meant by this one? @fickludd

Comment (fickludd) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: I just though maybe this query could become even shorter:

```
OPTIONAL MATCH (req:Y)
WITH req
OPTIONAL MATCH (req)<-[:*2]-(y)
RETURN DISTINCT req.eid, y.eid
```

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 2 above)
Comment (fickludd) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: So do we need this test then, if it adds no information? The reason I'm fretting is that this test is sort of hard to understand to me at least: it has two quite complex setup queries, and a quite big test query, all with labels and types and var-length patterns. When this test fails, future me will have to parse through all of that to be able to know what could cause the failure. If it does not expose any known problem that the short test cannot expose, it is much easier to not have to read it =). 

I'm open to be convinced though!

**Pull request merged**